### ISSUE TITLE
[CLI]: define_metric for step_metric fails if a later step is out of sync with the non step_metric step

### ISSUE THREAD

**Issue by rockerBOO** [wandb/wandb#6554]
### Describe the bug

If we have 2 different steps metrics. Like for training we use the default metric and a validation pass in the epoch has a different metric it fails if the validation steps are greater than the default metric.

```python
import wandb
import random

wandb.login()

random.seed(1)

# Initalize a new run
wandb.init(project="define-metric-demo", notes="custom step")

# Define the custom x axis metric
wandb.define_metric("custom_step")

# Define which metrics to plot against that x-axis
wandb.define_metric("validation/loss", step_metric="custom_step")

global_step = 0
for epoch in range(10):
    for i in range(3): # note this is smaller than the second range for "validation"
        global_step += 1
        log_dict = {
            "train/loss": 1 / (i + 1),
            "custom_step": i**2,
        }
        wandb.log(log_dict, step=global_step)

    for i in range(10):
        log_dict = {"validation/loss": 1 / (i + 1)}
        wandb.log(log_dict, step=(len(range(10)) * (epoch)) + 1 + i)

wandb.finish()
```

```shell
wandb: Currently logged in as: rockerboo. Use `wandb login --relogin` to force relogin
wandb: Tracking run with wandb version 0.15.12
wandb: Run data is saved locally in /mnt/900/builds/sd-scripts/wandb/run-20231105_153116-a0ru7zfd
wandb: Run `wandb offline` to turn off syncing.
wandb: Syncing run rural-spaceship-10
wandb: ⭐️ View project at https://wandb.ai/rockerboo/define-metric-demo
wandb: 🚀 View run at https://wandb.ai/rockerboo/define-metric-demo/runs/a0ru7zfd
wandb: Waiting for W&B process to finish... (success).
wandb: 
wandb: Run history:
wandb:     custom_step ▁███████████████████████████████████████
wandb:      train/loss █▃▁
wandb: validation/loss ▅▃▂▁█▃▂▁█▃▂▁█▃▂▁█▃▂▁█▃▂▁█▃▂▁█▃▂▁█▃▂▁█▃▂▁
wandb: 
wandb: Run summary:
wandb:     custom_step 4
wandb:      train/loss 0.33333
wandb: validation/loss 0.1
wandb: 
wandb: 🚀 View run rural-spaceship-10 at: https://wandb.ai/rockerboo/define-metric-demo/runs/a0ru7zfd
wandb: ️⚡ View job at https://wandb.ai/rockerboo/define-metric-demo/jobs/QXJ0aWZhY3RDb2xsZWN0aW9uOjExMjkyMTQ0MQ==/version_details/v2
wandb: Synced 4 W&B file(s), 0 media file(s), 0 artifact file(s) and 0 other file(s)
wandb: Find logs at: ./wandb/run-20231105_153116-a0ru7zfd/logs
wandb: WARNING Step only supports monotonically increasing values, use define_metric to set a custom x axis. For details see: https://wand
b.me/define-metric
wandb: WARNING (User provided step: 1 is less than current step: 3. Dropping entry: {'validation/loss': 1.0, '_timestamp': 1699216279.1592
035}).
wandb: WARNING (User provided step: 2 is less than current step: 3. Dropping entry: {'validation/loss': 0.5, '_timestamp': 1699216279.1592
505}).
wandb: WARNING (User provided step: 4 is less than current step: 10. Dropping entry: {'train/loss': 1.0, 'custom_step': 0, '_timestamp': 1
699216279.159862}).
wandb: WARNING (User provided step: 5 is less than current step: 10. Dropping entry: {'train/loss': 0.5, 'custom_step': 1, '_timestamp': 1
699216279.1599138}).
wandb: WARNING (User provided step: 6 is less than current step: 10. Dropping entry: {'train/loss': 0.3333333333333333, 'custom_step': 4, 
'_timestamp': 1699216279.1599715}).
wandb: WARNING (User provided step: 7 is less than current step: 20. Dropping entry: {'train/loss': 1.0, 'custom_step': 0, '_timestamp': 1
699216279.1604714}).
wandb: WARNING (User provided step: 8 is less than current step: 20. Dropping entry: {'train/loss': 0.5, 'custom_step': 1, '_timestamp': 1
699216279.1605194}).
wandb: WARNING (User provided step: 9 is less than current step: 20. Dropping entry: {'train/loss': 0.3333333333333333, 'custom_step': 4, 
'_timestamp': 1699216279.1605668}).
wandb: WARNING (User provided step: 10 is less than current step: 30. Dropping entry: {'train/loss': 1.0, 'custom_step': 0, '_timestamp': 
1699216279.1610472}).
wandb: WARNING (User provided step: 11 is less than current step: 30. Dropping entry: {'train/loss': 0.5, 'custom_step': 1, '_timestamp': 
1699216279.1610982}).
wandb: WARNING (User provided step: 12 is less than current step: 30. Dropping entry: {'train/loss': 0.3333333333333333, 'custom_step': 4,
 '_timestamp': 1699216279.1611462}).
wandb: WARNING (User provided step: 13 is less than current step: 40. Dropping entry: {'train/loss': 1.0, 'custom_step': 0, '_timestamp': 
1699216279.161626}).
wandb: WARNING (User provided step: 14 is less than current step: 40. Dropping entry: {'train/loss': 0.5, 'custom_step': 1, '_timestamp': 
1699216279.1616735}).
wandb: WARNING (User provided step: 15 is less than current step: 40. Dropping entry: {'train/loss': 0.3333333333333333, 'custom_step': 4,
 '_timestamp': 1699216279.161721}).
wandb: WARNING (User provided step: 16 is less than current step: 50. Dropping entry: {'train/loss': 1.0, 'custom_step': 0, '_timestamp': 
1699216279.162198}).
wandb: WARNING (User provided step: 17 is less than current step: 50. Dropping entry: {'train/loss': 0.5, 'custom_step': 1, '_timestamp': 
1699216279.162246}).
wandb: WARNING (User provided step: 18 is less than current step: 50. Dropping entry: {'train/loss': 0.3333333333333333, 'custom_step': 4,
 '_timestamp': 1699216279.162293}).
wandb: WARNING (User provided step: 19 is less than current step: 60. Dropping entry: {'train/loss': 1.0, 'custom_step': 0, '_timestamp': 
1699216279.1627688}).
wandb: WARNING (User provided step: 20 is less than current step: 60. Dropping entry: {'train/loss': 0.5, 'custom_step': 1, '_timestamp': 
1699216279.162816}).
wandb: WARNING (User provided step: 21 is less than current step: 60. Dropping entry: {'train/loss': 0.3333333333333333, 'custom_step': 4,
 '_timestamp': 1699216279.162863}).
wandb: WARNING (User provided step: 22 is less than current step: 70. Dropping entry: {'train/loss': 1.0, 'custom_step': 0, '_timestamp': 
1699216279.1633492}).
wandb: WARNING (User provided step: 23 is less than current step: 70. Dropping entry: {'train/loss': 0.5, 'custom_step': 1, '_timestamp': 
1699216279.1633968}).
wandb: WARNING (User provided step: 24 is less than current step: 70. Dropping entry: {'train/loss': 0.3333333333333333, 'custom_step': 4,
 '_timestamp': 1699216279.1634438}).
wandb: WARNING (User provided step: 25 is less than current step: 80. Dropping entry: {'train/loss': 1.0, 'custom_step': 0, '_timestamp': 
1699216279.1639175}).
wandb: WARNING (User provided step: 26 is less than current step: 80. Dropping entry: {'train/loss': 0.5, 'custom_step': 1, '_timestamp': 
1699216279.1639645}).
wandb: WARNING (User provided step: 27 is less than current step: 80. Dropping entry: {'train/loss': 0.3333333333333333, 'custom_step': 4,
 '_timestamp': 1699216279.1640112}).
wandb: WARNING (User provided step: 28 is less than current step: 90. Dropping entry: {'train/loss': 1.0, 'custom_step': 0, '_timestamp': 
1699216279.164489}).
wandb: WARNING (User provided step: 29 is less than current step: 90. Dropping entry: {'train/loss': 0.5, 'custom_step': 1, '_timestamp': 
1699216279.164536}).
wandb: WARNING (User provided step: 30 is less than current step: 90. Dropping entry: {'train/loss': 0.3333333333333333, 'custom_step': 4,
 '_timestamp': 1699216279.1645827}).
```

### Environment

WandB version: 0.15.12

OS: arch linux 

Python version: 3.10.11

Versions of relevant libraries:

### Additional Context

I'm expecting that if I am using a separate step metric that it wouldn't conflict with the default steps. This specific problem occurs if I'm using a gradient accumulation steps that is large or full batch size and using stochastic batch size on the validation dataset. (3 steps for training, 10 steps for validation)

Thank you!

**Comment by MBakirWB** [comment]
Hi @rockerBOO , thanks for writing in. We will review the provide material and get back to you. Thanks

**Comment by MBakirWB** [comment]
Hi @rockerBOO , the reason you are seeing this error is you are incrementing your custom axis value in your first loop ` "custom_step": i**2,`. The same step incrementation rules apply to your custom defined metric, it must also monotonically increase. In subsequent loops for you validation/loss, you begin trying to log to earlier steps since validation/loss is defined against the custom_step when custom_step has already incremented beyond your calculated value in `step=(len(range(10)) * (epoch)) + 1 + i`

**Comment by rockerBOO** [comment]
Thank you @MBakirWB Just got some time to review. I'm more confused now how the "custom_step" is getting the step metric that is defined separately and not assigned to that value as a step_metric like validation/loss is set. To clear this up I have tried this example where it doesn't use the same name.

```python
import wandb
import random

wandb.login()

random.seed(1)

# initalize a new run
wandb.init(project="define-metric-demo", notes="validation step check")

# define the custom x axis metric
wandb.define_metric("validation_loss_step")

# define which metrics to plot against that x-axis
wandb.define_metric("validation/loss", step_metric="validation_loss_step")

global_step = 0
for epoch in range(10):
    for i in range(3):
        global_step += 1
        log_dict = {
            "train/loss": 1 / (i + 1),
            "custom_step": i**2,
        }
        wandb.log(log_dict, step=global_step)

    for i in range(10):
        log_dict = {"validation/loss": 1 / (i + 1)}
        wandb.log(log_dict, step=(len(range(10)) * (epoch)) + 1 + i)

wandb.finish()
```

```
wandb: WARNING Step only supports monotonically increasing values, use define_metric to set a custom x axis. For details see: https://wandb.me/define-metric
wandb: WARNING (User provided step: 1 is less than current step: 3. Dropping entry: {'validation/loss': 1.0, '_timestamp': 1700446586.5844166}).
wandb: WARNING (User provided step: 2 is less than current step: 3. Dropping entry: {'validation/loss': 0.5, '_timestamp': 1700446586.584464}).
wandb: WARNING (User provided step: 4 is less than current step: 10. Dropping entry: {'train/loss': 1.0, 'custom_step': 0, '_timestamp': 1700446586.58488}).
wandb: WARNING (User provided step: 5 is less than current step: 10. Dropping entry: {'train/loss': 0.5, 'custom_step': 1, '_timestamp': 1700446586.5849304}).
wandb: WARNING (User provided step: 6 is less than current step: 10. Dropping entry: {'train/loss': 0.3333333333333333, 'custom_step': 4, '_timestamp':
1700446586.5849924}).
wandb: WARNING (User provided step: 7 is less than current step: 20. Dropping entry: {'train/loss': 1.0, 'custom_step': 0, '_timestamp': 1700446586.5855734}).
...
```

This was working before and when updating to some newer version caused this warning to show up. Unfortunately this warning causes it to not be able to record or graph any of these results. If I am doing this inaccurately, how do I fix this? I tried every combo of define_metric and step_metric I could think of and read over the https://wandb.me/define-metric. I am still at a loss. 

Thank you for looking into this. Sorry I am a little frustrated.

**Comment by MBakirWB** [comment]
Hi @rockerBOO , 

Thank you for the additional feedback. Please note we've introduced this warning `wandb: WARNING (User provided step:...` to inform users if they are inadvertently logging incorrectly. See notes below to update your code

```
import wandb
import random
wandb.login()
random.seed(1)

# initalize a new run
wandb.init(project="define-metric-demo", notes="validation step check")

# Note a metric to be used as a step metric only
wandb.define_metric("validation_loss_step")  

# Associating the step metric with the validation/loss
wandb.define_metric("validation/loss", step_metric="validation_loss_step")  

global_step = 0
for epoch in range(10):
    for i in range(3):
        global_step += 1
        log_dict = {
            "train/loss": 1 / (i + 1),
            "custom_step": i**2,
        }
        wandb.log(log_dict) #remove step argument. These wandb calls will monotonically increase against the internal wandb step, so by last epoch, you will have logged 30 values for "train/loss" & "custom_step"

    for i in range(10):
        log_dict = {"validation_loss_step": len(range(10)) * (epoch) + 1 + i, #here you are explicitly incrementing the step value of your defined metric you want to log validation/loss against
                    "validation/loss": i #here i am logging a value for my metric that is being logged against my custom access
                    }
        wandb.log(log_dict)

wandb.finish()
```

**Comment by rockerBOO** [comment]
Ohhh, I see what this all means now... 

To clarify my confusion. There is a "step=" property in the log component that hints this is where you should set the steps. But your clarification indicates this ONLY should be used to modify the global step. "Custom" steps must be defined as a specific metric and logged using the log dict. This makes sense now. 

I think a further clarification in the documentation to hint away from setting the step= value to set the custom steps. Maybe further explaining what "custom_step" is how you define the step metric value. After getting your clarification it seems straight forward in reading the define-metrics page but this intuition mismatch might be difficult to work through. 

Maybe just another example using different terms could be enough to explain it with enough variation to see what the patterns should be. Your example showing a workspace was also good so I could see it more clearly but we also were on the same page of what I was expecting. I just didn't get that I needed to add the step in the log dict.

Thank you for taking the time to walk through this with me and helping me figure this out.
